### ISSUE TITLE
Create a Smart equivalent to React Flow's Step Edge

### ISSUE THREAD

**Issue by alexey-koren** [tisoap/react-flow-smart-edge#12]
Is it possible to support Step Edge type from react-flow?

![image](https://user-images.githubusercontent.com/36141306/153771168-d6354900-965b-4475-9697-08bb9ad911de.png)

**Comment by tisoap** [comment]
@alexey-koren Could you expand on that? What do you mean by "support Step Edge"?

**Comment by alexey-koren** [comment]
> @alexey-koren Could you expand on that? What do you mean by "support Step Edge"?

Yes. Now u have 2 types of lineType, 'curve' and 'straight' , or your custom edge path

I meant to support 'step' edge lineType from https://reactflow.dev/examples/edges/ (example edge from node 2 to node 3)
And add new lineType to SmartEdgeProvider options

**Comment by tisoap** [comment]
Ah, I see what you mean. Right now the `curve` and `straight` line types are analogous to the native `Bezier Edge` and `Straight Edge` from React Flow, and you're asking for something analogous to `Step Edge` or `Smooth Step Edge`.

The problem with implement a `step` line type is that it's no longer the shortest path from two connections, so I can't simply feed the bounding boxes to a path finding algorithm and be done, it would require an entire new implementation. 

I'm using the [`pathfinding` module](https://www.npmjs.com/package/pathfinding) behind the scenes, and even setting `allowDiagonal` to `false` [still creates some diagonal paths](https://tisoap.github.io/react-flow-smart-edge/?path=/story/smartedge--straight-lines-with-less-corners), and I'm not sure why, maybe it has something to do with the way the XY coordinates are being converted to the input matrix.

Long story short, it should be possible but it'll be a lot of work and I don't know when I'll have the time to do so.

**Comment by tisoap** [comment]
@alexey-koren  I remembered a user on the discord server came up with a working implementation of an "step" edge that wraps around nodes, but uses hard coded values. [Here's the direct link](https://discord.com/channels/771389069270712320/859801398228156446/902562586879479909)

**Comment by alexey-koren** [comment]
@tisoap Hi. Thank you for your response. We did it by ourselves with JumpPointFinder from pathfinding, and fullPath.
![image](https://user-images.githubusercontent.com/36141306/155313250-652c9a67-5e82-4c53-ba85-0377c3456955.png)

**Comment by alexey-koren** [comment]
@tisoap I can try to add this to your library, if needs

**Comment by tisoap** [comment]
@alexey-koren thanks! But I'm actually waiting for the [v10 release](https://github.com/wbkd/react-flow/pull/1555) of React Flow before updating stuff, because I don't want to re-do work to make it compatible with the new version. I also want to introduce a breaking change on how `SmartEdge` is configured to allow for [multiple types to be used at the same time](https://github.com/tisoap/react-flow-smart-edge/issues/11), so I wouldn't worry about adding a smart `step` edge so soon.

But I'll definitely look into `JumpPointFinder`, thanks for mentioning it! I'll leave this issue open as a reminder for me

**Comment by lichenglu** [comment]
> @tisoap Hi. Thank you for your response. We did it by ourselves with JumpPointFinder from pathfinding, and fullPath. ![SmartEdge](https://user-images.githubusercontent.com/36141306/155313848-347927fd-ca42-4a1f-a368-f7fcc20bba7b.gif)

This looks awesome! Could share more details on how you implemented it? Thanks!

**Comment by StringKe** [comment]
> @tisoap Hi. Thank you for your response. We did it by ourselves with JumpPointFinder from pathfinding, and fullPath. ![SmartEdge](https://user-images.githubusercontent.com/36141306/155313848-347927fd-ca42-4a1f-a368-f7fcc20bba7b.gif)

This works great, I need to implement something similar, can you tell me how you did it?

I'm currently trying to handle it with the collision algorithm sat.js but it doesn't work well.

**Comment by liljefelt** [comment]
> > @tisoap Hi. Thank you for your response. We did it by ourselves with JumpPointFinder from pathfinding, and fullPath. ![SmartEdge](https://user-images.githubusercontent.com/36141306/155313848-347927fd-ca42-4a1f-a368-f7fcc20bba7b.gif)
> 
> This works great, I need to implement something similar, can you tell me how you did it?
> 
> I'm currently trying to handle it with the collision algorithm sat.js but it doesn't work well.

I have nothing to add but I'm also interested!

**Comment by alexey-koren** [comment]
Hi guys, sorry, I'm was in Ukraine, so wasn't available
I can answer your all questions tomorrow

**Comment by StringKe** [comment]
@alexey-koren I'm guessing you should be using a pathfinding algorithm like `astar` here?

**Comment by alexey-koren** [comment]
> @alexey-koren I'm guessing you should be using a pathfinding algorithm like `astar` here?

Hi. I'm added example to gist https://gist.github.com/alexey-koren/d15f1e4103894fadbbe585f53d93e29e

**Comment by tisoap** [comment]
Hey everyone! (cc @alexey-koren @lichenglu @StringKe @liljefelt )

I've released a beta version that includes a `SmartStepEdge` export, could you guys test it?

Available to test using the `beta` tag:
```
npm install @tisoap/react-flow-smart-edge@beta
```

New documentation can be found [here](https://github.com/tisoap/react-flow-smart-edge/tree/feat/factory-function#readme)
